### ISSUE TITLE
[firebase_admob] ads invades safe area and close button is blocked on iPhone X

### ISSUE THREAD

**Issue by longphanmn** [FirebaseExtended/flutterfire#3090]
**Describe the bug**
Some of the interstitial ads are not rendered right on iPhone X, the close button is blocked.

**To Reproduce**
1. Create project with firebase_admob 0.9.3+4 and pod 'Google-Mobile-Ads-SDK' ~> '7.63.0'
2. Build and install release app on iPhone X 13.6
3. Some of the interstitial ads are not rendered right on iPhone X, the close button is blocked. Mostly games.

**Expected behavior**
Ad displayed within safe area

**Additional context**
Screenshots:

![ảnh](https://user-images.githubusercontent.com/21080279/89284762-74f7b880-d679-11ea-8a4c-b1e307f9589f.png)
![ảnh-2](https://user-images.githubusercontent.com/21080279/89284774-7923d600-d679-11ea-8ace-713359dab7ac.png)
![ảnh-3](https://user-images.githubusercontent.com/21080279/89284814-8640c500-d679-11ea-9e9e-b9d6493cd837.png)

**Comment by belisqui** [comment]
I’m having the exact same problem. They closed my ticked as you had already reported it.

**Comment by blasten** [comment]
This doesn't seem right. This is a duplicate of https://github.com/FirebaseExtended/flutterfire/issues/2531.

@jjliu15 how is this handled in the AdMob SDK?

**Comment by longphanmn** [comment]
> This doesn't seem right. This is a duplicate of #2531.
> 
> @jjliu15 how is this handled in the AdMob SDK?

I've never had this issue on rewarded ads. Only in interstitial.
#2531 is only rewarded ads, look like many fullscreen ads has this issue.

**Comment by code-kooc** [comment]
I have same issue in interstitial ad.

**Comment by wonpyohong** [comment]
My app users complain to me about this.
This is severe because user should close app completely and start app again.
This is terrible User Experience

**Comment by wonpyohong** [comment]
It can be reproduced simply test ads and simulator (iPhone SE, non-notch)

<img width="405" alt="스크린샷 2020-10-04 오전 3 26 45" src="https://user-images.githubusercontent.com/10491866/94999075-c678e100-05f1-11eb-99f4-41da34064226.png">

Upper half of close button is overlapped with status bar,
Lower half is not overlapped with status bar.

Upper half cannot be touched.
Close button should not be overlapped with status bar (like in notch style)

**Comment by summonerriftofficial** [comment]
Same here :(

**Comment by summonerriftofficial** [comment]
Many 1 star due to this lroblem, we need a fix

**Comment by Tibbs** [comment]
Is there any update on this? I had to disable the interstitial ads on most of my apps? I can't believe it is not a high priority issue.

**Comment by avbk** [comment]
I think this problem ist not only relevant on iOS, but on any device with rounded corner. For example I can reproduce this behaviour on a Google Pixel 4 as well.

Looking into the code the native part only invokes the show method of the InterstialAd class. I guess one cannot do much here, since the layout happens inside of the AdMob Library?

**Comment by talp-kzencorp** [comment]
This issue is causing horrible user experience.
Is this issue a bug with the AdMob iOS implementation?

**Comment by diegocom** [comment]
Same problem here. We need a solution

**Comment by Tibbs** [comment]
I found a workaround. If you hide the status bar in the presenting viewcontroller before the interstitial is presented, it obviously won't cover the close button.
You can hide the status bar by overwriting preferStatusbarHidden method

```
- (BOOL) prefersStatusBarHidden {
    return  hideStatusbar;
}

hideStatusbar = true;
[self prefersStatusBarHidden];
[self.admMobInterstitial presentFromRootViewController:self];

- (void) interstitialDidDismissScreen:(GADInterstitial *)interstitial {
    hideStatusbar = false;
    [self prefersStatusBarHidden];
}

- (void) interstitialWillLeaveApplication:(GADInterstitial *)ad {
    hideStatusbar = false;
    [self prefersStatusBarHidden];
}
```





If you don't have full control from which viewcontroller it's presented from 
It is a bit of a hack using the setStatusBarHidden deprecated API:

Before you present the interstitial, save the prefersStatusBarHidden value for the viewcontroller
```
statusBarHiddenForPresentingRootVC = rootVC.prefersStatusBarHidden;
[self.admMobInterstitial presentFromRootViewController:rootVC];

- (void) interstitialWillPresentScreen:(GADInterstitial *)ad {
    if (statusBarHiddenForPresentingRootVC == false) {
        [[UIApplication sharedApplication] setStatusBarHidden:YES];
    }
}

```

When the interstitial is dismissed or tapped, restore the statusbarHidden value

```
- (void) interstitialDidDismissScreen:(GADInterstitial *)interstitial {
    if (statusBarHiddenForPresentingRootVC == false) {
        [[UIApplication sharedApplication] setStatusBarHidden:false];
    }
}

- (void) interstitialWillLeaveApplication:(GADInterstitial *)ad {
    if (statusBarHiddenForPresentingRootVC == false) {
        [[UIApplication sharedApplication] setStatusBarHidden:false];
    }
}

```

**Comment by fleoparra** [comment]
Same problem here. I'm using firebase_admob: ^0.10.2. Close Button do not work.

![Simulator Screen Shot - iPhone 12 Pro Max - 2020-11-07 at 22 44 25](https://user-images.githubusercontent.com/4581151/98456369-e2205b80-214a-11eb-9298-237ca740db23.png)

**Comment by csongorkeller** [comment]
hiding the statusbar fixes for me, however it's not the best way, but works:
SystemChrome.setEnabledSystemUIOverlays([]);

**Comment by shunnokw** [comment]
Temporary solution:
Hide status bar when ads are shown
`    listener: (MobileAdEvent event) {
      if (event == MobileAdEvent.closed) {
        SystemChrome.setEnabledSystemUIOverlays(SystemUiOverlay.values);
      }
      if (event == MobileAdEvent.clicked) {
        SystemChrome.setEnabledSystemUIOverlays([]);
      }
    },`

**Comment by clk0115** [comment]
"Hide status bar" is not work fo me=[.
 Checked iPhone 12/ iPhone 12 Pro / iPhone 12 Pro Max  in iOS 14.1. The close button is under status bar. Can't close the ads.
But, except the about models, every model work prefect including iPhone 11 Pro Max in iOS 14.1.

**Comment by erickcchoi** [comment]
Thank you for the solution, Tested with real device iPhone 12 Pro Max : WORK

**Comment by AliRezaeian** [comment]
One issue is I added this ads on an Exit button where I pop the page. So in this case since I pop then the if (event == MobileAdEvent.closed) condition will never trigger. So the status bar will be hidden and not be displayed again after ads is closed. Any solution?

**Comment by Kenoshen** [comment]
Same issue.  Any update?

**Comment by erperejildo** [comment]
> Same issue. Any update?

This was my approach:
- Remove this line from styles.xml:
`<item name="android:windowFullscreen">true</item>`
- If you need to have the Android navigations button hide only the status bar:
`SystemChrome.setEnabledSystemUIOverlays([SystemUiOverlay.bottom]);`

The close button for the interstitial won't be an issue because it's on top. Same thing with the banner if you have it on top

**Comment by Vardiak** [comment]
This issue makes this plugin unusable in production. Why is it still not fixed after 5 months ?

**Comment by mulderpf** [comment]
Ok, so yet another thing to work around in the AdMob library (been using banner ads for two years now with a workaround in place because they will just overlay everything if you don't) and I just implemented interstitials in my app (since AdMob's been pestering me for ages to do so) and now I find that I cannot close the ad on iOS.  I'm also wondering if I will run into issue https://github.com/FirebaseExtended/flutterfire/issues/63 as I am using this together with banner ads.  I've been told for over a year now that this library is being improved, but it really doesn't feel like it is?

**Comment by rosaroko** [comment]
Ok, I encountered the same issues quite recently but without Flutter and I thought it might help you guys. I used @Tibbs workaround for my solution, so dude thank you! I didn't want to use any of the deprecated apis so I passed two callbacks from a ViewController to the GADInterstitialDelegate which are hides and shows the statusbar. It is not the most elegant solution but at least it works until it is fixed. 

For that to work you need to disable "View controller-based status bar appearance" in your Info.plist.

Go to your viewController which shows the ad define a global variable:
`BOOL statusBarHidden;`

set the it to set `YES` during the `viewDidLoad` and implement these three methods: 
```
- (void)hideStatusBar{
    statusBarHidden = YES;
    [self setNeedsStatusBarAppearanceUpdate];
}

 - (void)showStatusBar{
    statusBarHidden = NO;
    [self setNeedsStatusBarAppearanceUpdate];
}

- (BOOL)prefersStatusBarHidden {
    return statusBarHidden;
}
```

Go to your `GADInterstitialDelegate` and define there two callbacks (in my case they are in Swift) as global variables: 
```
 @objc var showStatusBarCallback: () -> Void = {}
 @objc var hideStatusBarCallback: () -> Void = {}
```

Implement the delegate methods:
 ```
func interstitialWillPresentScreen(_ ad: GADInterstitial) {
        self.hideStatusBarCallback()
}
    
func interstitial(_ ad: GADInterstitial, didFailToReceiveAdWithError error: GADRequestError) {
        self.showStatusBarCallback()
}
    
func interstitialDidDismissScreen(_ ad: GADInterstitial) {
        self.showStatusBarCallback()
}
    
func interstitialWillLeaveApplication(_ ad: GADInterstitial) {
        self.showStatusBarCallback()
}
```
you can pass these two callbacks during the initialization of the GADInterstitialDelegate and they will call hide and show statusbar every time it opens or closes the interstitial ad. 
Additionally you might need to set UIStatusBarStyle to the ViewController or a parent ViewController.

**Comment by belisqui** [comment]
Just received another complaint. “I can get past the ad it just gets stuck on it and won’t let me go forward” 
Any solution for flutter guys?

**Comment by memmen** [comment]
I am currently having the same issue. The x button is hidden in the upper left.

**Comment by Marek00Malik** [comment]
Any updates on this task ? The SystemChrome.setEnabledSystemUIOverlays is not working for all ads. If I get a full screen video, the x button is displayed behind the clock at the top left corner and I’m not able to close even the add after it’s finished.  Making the user experience horribly....

**Comment by russellwheatley** [comment]
The `firebase_admob` plugin will be deprecated in April 2021 in favor of [Google Mobile Ads SDK for Flutter](https://github.com/googleads/googleads-mobile-flutter).
Google Mobile Ads SDK for Flutter is a new Flutter plugin that supports more Ads formats than `firebase_admob` does. Google Mobile Ads SDK for Flutter currently supports loading and displaying banner, interstitial (full-screen), native ads, and rewarded video ads across AdMob and AdManager. It also supports displaying banner and native ads as `Widgets` as opposed to being overlayed over all app content.
We're now recommending projects currently using `firebase_admob` migrate to Google Mobile Ads SDK for Flutter following the [instructions outlined here](https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_admob#migrating-to-google-mobile-ads-sdk-for-flutter).
Issues with migrating or with using the new package should now be raised over on the [Google Mobile Ads SDK for Flutter](https://github.com/googleads/googleads-mobile-flutter) repository.
